src/storage/observer_worker.rs

```diff
@@ -99,56 +97,62 @@ where
         }
     }
 
-    async fn try_update(&self) -> Result<()> {
-        trace!("try update active blob");
-        let inner_cloned = self.inner.clone();
-        if let Some(inner) = active_blob_check(inner_cloned).await? {
-            update_active_blob(inner.clone()).await?;
-            inner.try_dump_old_blob_indexes(self.dump_sem.clone()).await;
+    async fn try_update_active_blob(&self) -> Result<bool> {
+        let config_max_size = self
+            .inner
+            .config
+            .max_blob_size()
+            .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
+        let config_max_count = self
+            .inner
+            .config
+            .max_data_in_blob()
+            .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
+
+        {
+            let read = self.inner.safe.read().await;
+            let active_blob = read.active_blob.as_ref();
+            if let Some(active_blob) = active_blob {
+                if active_blob.file_size() < config_max_size
+                    && (active_blob.records_count() as u64) < config_max_count
+                {
+                    return Ok(false);
+                }
+            }
         }
-        Ok(())
-    }
-}
 
-async fn active_blob_check<K>(inner: Inner<K>) -> Result<Option<Inner<K>>>
-where
-    for<'a> K: Key<'a> + 'static,
-{
-    let (active_size, active_count) = {
-        trace!("await for lock");
-        let safe_locked = inner.safe.read().await;
-        trace!("lock acquired");
-        if let Some(active_blob) = safe_locked.active_blob.as_ref() {
-            (active_blob.file_size(), active_blob.records_count() as u64)
-        } else {
-            // if active blob doesn't exists, it doesn't need to be updated
-            return Ok(None);
+        let mut write = self.inner.safe.write().await;
+        let active_blob = write.active_blob.as_ref();
+        if let Some(active_blob) = active_blob {
+            if active_blob.file_size() >= config_max_size
+                || active_blob.records_count() as u64 >= config_max_count
+            {
+                let next_name = self.inner.next_blob_name()?;

```

Comment:
You have `update_active_blob` function below. Why not call it from here?

Reply:
We have lock already taken here, and update function takes lock later

Reply:
It also means more actions during file lock (obtaining name and creating actual blob file) than before. It does not create any noticeable overhead during tests.

Reply:
Oh, we don't need complex lock logic here since it can be accessed from single thread only. You're right, update blob can be reused